docs/proposals/20210203-unmanaged-infrastructure.md

```diff
@@ -0,0 +1,219 @@
+The machine controller must be able to operate without hard dependencies regardless of the cluster infrastructure being managed or externally managed.
+![](https://i.imgur.com/nA61XJt.png)
+
+### User Stories
+
+#### Story 1 - Alternate control plane provisioning with user managed infrastructure
+As a cluster provider I want to use CAPI in my service offering to orchestrate Kubernetes bootstraping while letting workload cluster operators own their infrastructure lifecycle.
+
+#### Story 2 - Restricted access to cloud provider APIs
+As a cluster operator I want to use CAPI to orchestrate kubernetes bootstraping while restricting the privileges I need to grant for my cloud provider because of organisational cloud security constraints.
+
+#### Story 3 - Consuming existing cloud infrastructure
+As a cluster operator I want to use CAPI to orchestate Kubernetes bootstraping while reusing infrastructure that has already been created in the organisation either by me or another team.
+
+### Implementation Details/Notes/Constraints
+
+**Managed**
+- It will be default and will preserve existing behaviour.
+
+**Externally Managed**
+
+The provider InfraCluster controller will:
+- Skip any reconciliation of the resource.
+
+- Will not update the resource or its status in any way
+
+The external management system will:
+
+- Populate all required fields within the InfraCluster spec to allow other CAPI components to continue as normal.
+
+- When the infrastructure is ready, set the appropriate status as is done by the provider controller today.
+
+#### Provider implementation changes
+
+To enable providers to implement the changes required by this contract, we will provide a new `InfraClusterExternallyManaged` predicate.
+
+This will filter out any InfraCluster resource that has been marked as "externally managed" and will prevent the controller from reconciling the resource.
+
+### Security Model
+
+When externally managed, no additional privileges for a cloud provider need to be given to CAPI other than the required to manage machines.
+
+### Risks and Mitigations
+
+#### What happens when a user coverts an externally managed InfraCluster to a managed InfraCluster?
+
+Whether the flag for being managed is an annotation or field on the InfraCluster, there currently is no immutability support for CRD fields or annotations within the Kubernetes API.
+
+This means that, once a user has created their externally managed InfraCluster, they could at some point, update the annotation/field to make the InfraCluster appear to be managed.
+
+There is no way to predict what would happen in this scenario.
+The InfraCluster controller would start attempting to reconcile infrastructure that it did not create, and therefore, there may be assumptions it makes that mean it cannot manage this infrastructure.
+
+To prevent this, we will have to implement (in the InfraCluster webhook) a means to prevent users converting externally managed InfraClusters into managed InfraClusters.
```

Comment:
What about adding this as a requirement for the external management system up in the document?

Reply:
TBH I would prefer to lust this as a responsibility of the External providers, because I don't see how this can be centralised given that we are talking about InfraCluster objects and one of the of the goal of this KEP is to avoid changes to the existing infrastructure providers.
If instead you can come out with a solution that centralises this responsibility, +1 to that

Reply:
I was wondering about a generic validating webhook that could be implemented to look at the metadata only (therefore could accept any type), this code could be provided centrally and generic, but providers or external infrastructure providers would need to set it up to look at the correct type

Ie all they would need to do is configure the ValidatingWebhookConfiguration correctly. Do you think something like this would work/be useful?